### PR TITLE
Regenerated rest api client and Refactored workspace connection parameters

### PR DESCRIPTION
**Pull request by vxfield**
Replaces #558 and #564

## Main changes:
1. Regenerated the rest api client based on latest swagger
2. Refactored the parameters to connect to a workspace into its own class
3. Gathered common constants into a new _constants.py file
4. Cleaned up code/logic of the _authentication module
5. Cleaned up and fixed some bugs in the unit test framework
6. Refactored, cleaned up, and added more tests for the workspace connection parameters and authentication
7. Fixed many linting errors

### CODE REVIEW

File: azure-quantum/.vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (04diiguyi) [inline azure-quantum/.vscode/settings.json:1]: Do we need to add this file?

File: azure-quantum/azure/quantum/workspace.py [hunk 2]
```diff
@@ -132,110 +106,91 @@ def __init__(
...
         # Create QuantumClient
         self._client = self._create_client()
 
-    def _create_client(self) -> QuantumClient:
-        endpoint = BASE_URL(self.location)
-        logger.debug(
-            f"Creating client for: subs:{self.subscription_id},"
-            + f"rg={self.resource_group}, ws={self.name}, frontdoor={endpoint}"
-        )
+    def on_new_client_request(self):
```
Comment (04diiguyi) [inline azure-quantum/azure/quantum/workspace.py:133]: Why we need to create_client() to _client twice?

File: azure-quantum/.vscode/settings.json (hunk 1 above)
Comment (vxfield) [reply azure-quantum/.vscode/settings.json:1]: Having this file improves the SDK developer experience when running the unit tests from VS Code.
It's not really "needed", but nice to have :-)

File: azure-quantum/azure/quantum/workspace.py (hunk 2 above)
Comment (vxfield) [reply azure-quantum/azure/quantum/workspace.py:133]: We need to recreate the client when the user_agent changes.
This happens in the cirq and qiskit clients.

**Review by kikomiss — APPROVED** [review]
Thank you for cleaning up the `_DefaultAzureCredential` and `Workspace`! 

Since we now have auto-generated documentation for `azure-quantum` package we would need to pay attention what classes get added into it. 
I think the changed in this PR won't add any documentation for use as the `WorkspaceConnectionParams` is private and is not defined in `__all__` anywhere, but otherwise they would go into documentation.

**Comment by vxfield** [comment]
> Thank you for cleaning up the `_DefaultAzureCredential` and `Workspace`!
> 
> Since we now have auto-generated documentation for `azure-quantum` package we would need to pay attention what classes get added into it. I think the changed in this PR won't add any documentation for us as the `WorkspaceConnectionParams` is private and is not defined in `__all__` anywhere, but otherwise they would go into documentation.

Cool!
I kept it out of `__all__` for this reason.
Is there a way to run the doc-gen locally to see the results?

File: azure-quantum/azure/quantum/_constants.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    DATA_PLANE_CREDENTIAL_SCOPE = "https://quantum.microsoft.com/.default"
+    ARM_CREDENTIAL_SCOPE = "https://management.azure.com/.default"
+
+    MSA_TENANT_ID = "9188040d-6c67-4c5b-b112-36a304b66dad"
+
+    AUTHORITY = AzureIdentityInternals.get_default_authority()
+    DOGFOOD_AUTHORITY = "login.windows-ppe.net"
+
+    # pylint: disable=unnecessary-lambda-assignment
+    QUANTUM_BASE_URL = \
```
[the remark this answers is not in the document]
Comment (vxfield) [reply azure-quantum/azure/quantum/_constants.py:60]: nit comments -> neat code
I'll make changes to keep things more clear.

[the remark this answers is not in the document]
Comment (vxfield) [reply azure-quantum/azure/quantum/_constants.py]: Done.

**Comment by vxfield** [comment]
Superseded by #579

**Pull request closed without merging**